File: Design Patterns/Structural/Typescript/Composite.ts
// use for treating objects or group of object the same way  
interface ISmartDevice {
    turnOn(): void
    turnOff(): void
}

class Device implements ISmartDevice {
    constructor(private name: string) { }
    turnOn() { console.log(this.name + " turned ON") }
    turnOff() { console.log(this.name + " turned OFF") }
}

class DeviceGroup implements ISmartDevice {
    private devices: ISmartDevice[] = []

    constructor(private name: string) { }

    add(device: ISmartDevice) { this.devices.push(device) }

    turnOn() {
        console.log("Turning ON group: " + this.name)
        this.devices.forEach(d => d.turnOn())
    }

    turnOff() {
        console.log("Turning OFF group: " + this.name)
        this.devices.forEach(d => d.turnOff())
    }
}

const livingRoom = new DeviceGroup("livingRoom")
const TV = new Device("TV")
const Cleaner = new Device("Cleaner")
const CoffeeMaker = new Device("CoffeeMaker")

livingRoom.add(TV)
livingRoom.add(Cleaner)
livingRoom.add(CoffeeMaker)
livingRoom.turnOff()
TV.turnOn()

File: C#/Design Patterns/Structural/Composite.cs
using System;
using System.Collections.Generic;
namespace Composite
{
    interface ISmartDevice
    {
        void TurnOn();
        void TurnOff();
    }

    class Device : ISmartDevice
    {
        private string name;

        public Device(string name)
        { this.name = name; }

        public void TurnOn() { Console.WriteLine($"{name} turned ON"); }

        public void TurnOff() { Console.WriteLine($"{name} turned OFF"); }
    }

    class DeviceGroup : ISmartDevice
    {
        private string name;
        private List<ISmartDevice> devices = new List<ISmartDevice>();

        public DeviceGroup(string name) { this.name = name; }

        public void AddDevice(ISmartDevice device) { devices.Add(device); }

        public void TurnOn()
        {
            Console.WriteLine($"Turning ON group: {name}");
            foreach (var device in devices)
            { device.TurnOn(); }
        }

        public void TurnOff()
        {
            Console.WriteLine($"Turning OFF group: {name}");
            foreach (var device in devices)
            { device.TurnOff(); }
        }
    }

    class Program
    {
        public static void Main(string[] args)
        {
            DeviceGroup livingRoom = new DeviceGroup("livingRoom");
            Device TV = new Device("TV");
            Device Cleaner = new Device("Cleaner");
            Device CoffeeMaker = new Device("CoffeeMaker");

            livingRoom.AddDevice(TV);
            livingRoom.AddDevice(Cleaner);
            livingRoom.AddDevice(CoffeeMaker);
            livingRoom.TurnOff();
            TV.TurnOn();
        }
    }
}